### PR TITLE
REFACTOR: scroll-based mobile header switch

### PR DESCRIPTION
**Pull request by hnb-ku**
This refactor addresses the following issues:

1- Moves all relevant logic to the discourse-topic component (matches desktop)
2- Fixes the flicker issue discussed [here](https://meta.discourse.org/t/header-flickering-on-mobile/109359)
3- Fixes a rare occurring issue where tags wrap to a third line if a topic has long category names and lots of tags
4- Fixes header icon jitter on iOS
5- Fixes an issue where sliding out user / hamburger menus on Android leaves the user in a mid-state with half a title and the header panel visible - swiping will now open the menus but have no effect on the header.
6- adds min-width to the small-logo to act as placeholder so that the title doesn't shift if the logo takes a while to load.

Other than that, everything should look and act the same.

### CODE REVIEW

File: app/assets/javascripts/discourse/components/discourse-topic.js.es6 [hunk 1]
```diff
@@ -148,17 +131,13 @@ export default Ember.Component.extend(AddArchetypeClass, Scrolling, {
...
-      return true;
-      // condition for desktops
-    } else {
-      return offset > this.dockAt;
-    }
+    // On mobile, we show the header topic if the user has scrolled past the topic
+    // title and the current scroll direction is down
+    // On desktop the user only needs to scroll past the topic title.
+
+    return this.site.mobileView
```
Comment (nbianca) [inline app/assets/javascripts/discourse/components/discourse-topic.js.es6]: What about:

```suggestion
    return offset > this.dockAt && (!this.site.mobileView || this.mobileScrollDirection === "down")
```

File: app/assets/javascripts/discourse/components/discourse-topic.js.es6 [hunk 2]
```diff
@@ -193,25 +172,65 @@ export default Ember.Component.extend(AddArchetypeClass, Scrolling, {
...
-    // 3px buffer so that the switch doesn't happen with tiny scrolls
-    if (delta > 3 && scrollDirection !== "down") {
-      scrollDirection = "down";
-      this.appEvents.trigger("header:show-topic", this.topic);
-    } else if (delta < -3 && scrollDirection !== "up") {
-      scrollDirection = "up";
-      this.appEvents.trigger("header:hide-topic");
+  _mobileScrollDirectionCheck(offset) {
+    // Difference between this scroll and the one before it.
+    const delta = parseInt(offset - this._mobileLastScroll, 10);
```
Comment (nbianca) [inline app/assets/javascripts/discourse/components/discourse-topic.js.es6]: I believe you use `parseInt` here for type casting a float to int. What you could do is use `Math.floor`.

File: app/assets/javascripts/discourse/components/discourse-topic.js.es6 [hunk 3]
```diff
@@ -193,25 +172,65 @@ export default Ember.Component.extend(AddArchetypeClass, Scrolling, {
...
+      // Delta is positive so the direction is down
+      this.set("mobileScrollDirection", "down");
+    } else if (currDirection === "up" && prevDirection !== "up") {
+      // Delta is negative so the direction is up
+      this.set("mobileScrollDirection", "up");
     }
-    lastScroll = offset;
+
+    // We store this to compare against it the next time the user scrolls
+    this._mobileLastScroll = parseInt(offset, 10);
```
Comment (nbianca) [inline app/assets/javascripts/discourse/components/discourse-topic.js.es6]: Same as above.

File: app/assets/javascripts/discourse/components/discourse-topic.js.es6 [hunk 4]
```diff
@@ -193,25 +172,65 @@ export default Ember.Component.extend(AddArchetypeClass, Scrolling, {
...
+
+    // If the user reaches the very bottom of the topic, we want to reset the
+    // scroll direction in order for the header to switch back.
+    const distanceToTopicBottom = parseInt(
```
Comment (nbianca) [inline app/assets/javascripts/discourse/components/discourse-topic.js.es6]: Same as above.

File: app/assets/javascripts/discourse/components/discourse-topic.js.es6 [hunk 5]
```diff
@@ -193,25 +172,65 @@ export default Ember.Component.extend(AddArchetypeClass, Scrolling, {
...
+    const delta = parseInt(offset - this._mobileLastScroll, 10);
+
+    // This is a tiny scroll, so we ignore it.
+    if (delta <= MOBILE_SCROLL_TOLERANCE && delta >= -MOBILE_SCROLL_TOLERANCE)
+      return;
+
+    const prevDirection = this.mobileScrollDirection;
+    const currDirection = delta > 0 ? "down" : "up";
+
+    if (currDirection === "down" && prevDirection !== "down") {
```
Comment (nbianca) [inline app/assets/javascripts/discourse/components/discourse-topic.js.es6]: What about:

```
    if (currDirection !== prevDirection) {
      this.set("mobileScrollDirection", currDirection);
    }
```

File: app/assets/javascripts/discourse/components/discourse-topic.js.es6 [hunk 6]
```diff
@@ -193,25 +172,65 @@ export default Ember.Component.extend(AddArchetypeClass, Scrolling, {
...
+
+    // We're at the bottom now, so we reset the direction.
+    this.set("mobileScrollDirection", null);
+  },
+
+  // We observe the scroll direction on mobile and if it's down, we show the topic
+  // in the header, otherwise, we hide it.
+  @observes("mobileScrollDirection")
+  toggleMobileHeaderTopic() {
+    return this.mobileScrollDirection === "down"
```
Comment (nbianca) [inline app/assets/javascripts/discourse/components/discourse-topic.js.es6]:
```suggestion
    return this.appEvents.trigger("header:update-topic", this.mobileScrollDirection === "down" ? this.get("topic") : null);
```

**Comment by hnb-ku** [comment]
Thank you for the feedback and insights @nbianca! 

I've made the requested changes and will go ahead and merge this now.

**Pull request merged**